src/main/java/at/hannibal2/skyhanni/features/inventory/EssenceShopHelper.kt

```diff
@@ -0,0 +1,251 @@
+    private var currentEssenceItem: NEUInternalName? = null
+    private var ownedEssence: Int = 0
+    private var essenceNeeded: Int = 0
+    private var lastClick = SimpleTimeMark.farPast()
+
+    /**
+     * REGEX-TEST: Gold Essence Shop
+     * REGEX-TEST: Wither Essence Shop
+     */
+    private val essenceShopPattern by patternGroup.pattern(
+        "essence.shop",
+        "(?:§.)*(?<essence>.*) Essence Shop",
+    )
+
+    /**
+     * REGEX-TEST: §7Your Undead Essence: §d12,664
+     * REGEX-TEST: §7Your Wither Essence: §d2,275
+     */
+    private val currentEssenceCountPattern by patternGroup.pattern(
+        "essence.current",
+        ".*§7Your (?<essence>.*) Essence: §.(?<count>[\\d,]*)",
+    )
+
+
+    /**
+     * REGEX-TEST: §aHigh Roller I
+     * REGEX-TEST: §aForbidden Speed III
+     * REGEX-TEST: §aReturn to Sender X
+     */
+    private val essenceUpgradePattern by patternGroup.pattern(
+        "essence.upgrade",
+        "§.(?<upgrade>.*) (?<tier>[IVXLCDM]*)",
+    )
+
+    data class EssenceShop(val shopName: String, val upgrades: List<EssenceShopUpgrade>)
+    data class EssenceShopUpgradeStatus(
+        val upgradeName: String,
+        val currentLevel: Int,
+        val maxLevel: Int,
+        val remainingCosts: MutableList<Int>,
+    )
+
+    data class EssenceShopProgress(val essenceName: String, val purchasedUpgrades: Map<String, Int>) {
+        private val essenceShop = essenceShops.find { it.shopName.equals(essenceName, ignoreCase = true) }
+        val remainingUpgrades: MutableList<EssenceShopUpgradeStatus> = essenceShop?.upgrades?.map {
+            val purchasedAmount = purchasedUpgrades[it.name]
+            EssenceShopUpgradeStatus(
+                it.name,
+                currentLevel = purchasedAmount ?: 0,
+                maxLevel = it.costs.count(),
+                remainingCosts = (if (purchasedAmount == null) it.costs else it.costs.drop(purchasedAmount)).toMutableList(),
+            )
+        }?.toMutableList() ?: mutableListOf()
+        val nonRepoUpgrades = purchasedUpgrades.filter { purchasedUpgrade ->
+            essenceShop?.upgrades?.none { it.name.equals(purchasedUpgrade.key, ignoreCase = true) } == true
+        }
+    }
+
+    @SubscribeEvent
+    fun replaceItem(event: ReplaceItemEvent) {
+        if (!enabled() || essenceShops.isEmpty() || currentProgress == null) return
+        if (!essenceShopPattern.matches(event.inventory.name)) return
+        if (event.slot == CUSTOM_STACK_LOCATION) {
+            val progressStack = getCustomItemStack() ?: return
+            event.replace(progressStack)
+        }
+    }
+
+    @SubscribeEvent
+    fun onSlotClick(event: GuiContainerEvent.SlotClickEvent) {
+        if (currentProgress == null || event.slotId != CUSTOM_STACK_LOCATION || currentEssenceItem == null) return
+        event.cancel()
+        if (lastClick.passedSince() > 0.3.seconds) {
+            BazaarApi.searchForBazaarItem(currentEssenceItem!!, essenceNeeded)
```

Comment:
you can do `val currentEssenceItem = currentEssenceItem  ?: return` here so you dont need to do the double bang